### PR TITLE
Move semantics

### LINKED ISSUE #12: Move semantics or deep Copy by default?
Current design of C++ class wrappers around C types is causing double-free and invalid memory access, to elaborate, 
lets looks at  the IdlValue class:
```
class IdlValue {  
private:
    IDLValue *ptr;

public:
    // Create IdlValues from types
    IdlValue(IDLValue *ptr);
    IdlValue();
    explicit IdlValue(uint8_t val);
    explicit IdlValue(uint16_t val);
    explicit IdlValue(uint32_t val);
   // .. other members
   ~IdlValue() {
          if (ptr != nullptr)
                 idl_value_destroy(ptr);
   }
} 
```
The data member **ptr**, points to the inner **IDLValue** which was **allocated** in memory.
then, a user would consider safe to do the following:
```
auto elem = zondax::idl_value::IdlValue(nat64);
    {
        std::vector<zondax::idl_value::IdlValue> my_values;
        // push_back calls copy constructor of class IdlValue
        my_values.push_back(elem);
        // ... do other stuffs with my_values
        // finally my_values goes out of scope here, and IdlValue's destructor for each field is called 
    }
    .. // however here elem value goes out of scope, and destructor is called, then, 
    // it tries to freed the memory pointed by inner ptr, causing a **double-free** error 
     // and depending on the usage, invalid memory access.
```
this code as we can see, is not safe, because any copy of the original **elem**, would have a pointer pointing to the **same memory**, and each instance will try to **free** that memory once their destructor is called. see #10 and #9 
Posible ways to fix this:
1. Disable Copy in almost all C++ wrappers, enabling move semantics only, which is supported by C++.
2. Perform a deep copy, which implies additional memory allocations for any instance that is copied.
For the option 1:
```
auto value1 = IdlValue(10);
auto value2 = value2
```
will not compile, in order to do something similar:
`auto value2 = std::move(value2);`
value1 is left in an undefined state, following C++ standard and value2 takes ownership of inner pointer.

:link: [zboto Link](https://app.clickup.com/t/861mzaa22)

### PR DESCRIPTION
**Pull request by neithanmo**
This:
- enable move semantics in all classes to avoid double free errors, as inner pointer must have one owner at any time.
- Enhance file API to be more idiomatic.
- fix ic test issues, it is working now.

fixes #12 #9 and possible #10 

:link: [zboto Link](https://app.clickup.com/t/861mzankf)

### CODE REVIEW

File: lib-agent-cpp/inc/agent.h [hunk 1]
```diff
@@ -41,13 +41,26 @@ class Agent {
...
+        // declare move assignment
+        Agent& operator=(Agent &&o) noexcept;
+
         static std::variant<Agent, std::string> create_agent(std::string url, zondax::identity::Identity id,
             zondax::principal::Principal principal, const std::vector<char>& did_content); 
 
         ~Agent();
 
-        std::variant<IdlArgs, std::string> Query(std::string service, zondax::idl_args::IdlArgs args);
+        std::variant<IdlArgs, std::string> Query(std::string service, zondax::idl_args::IdlArgs &args);
```
Comment (ftheirs) [inline lib-agent-cpp/inc/agent.h:63]: For Query and Update functions, we need to check first that `agent` is not nullptr

File: lib-agent-cpp/inc/identity.h [hunk 2]
```diff
@@ -43,6 +43,16 @@ class Identity {
 public:
     ~Identity();
     Identity();
+
```
Comment (ftheirs) [inline lib-agent-cpp/inc/identity.h:46]: Similar to agent comment, please verify that `ptr` is not nullptr before calling rust functions. For instances, you can find an unsafe use in `Identity::Sign`

File: lib-agent-cpp/src/identity.cpp [hunk 3]
```diff
@@ -159,6 +184,7 @@ namespace zondax::identity {
         return type;
     }
 
+    // TODO: what function to call for identity destruction?
```
Comment (ftheirs) [inline lib-agent-cpp/src/identity.cpp:187]: Do we have a destruction function for this? Is this comment still relevant?

Comment (neithanmo) [reply lib-agent-cpp/src/identity.cpp]: still not.
related issue #16

Comment (chcmedeiros) [reply lib-agent-cpp/src/identity.cpp]: I will be working on that today/tomorrow. I can add on dev directly after the merge

Comment (neithanmo) [reply lib-agent-cpp/src/identity.cpp]: Depending on it we can decide if we remove moveSemantics for it, I am not sure this Identity class would hold a pointer to memory owned by rust.

Comment (neithanmo) [reply lib-agent-cpp/src/identity.cpp]: This type holds a pointer to memory allocated by rust.

File: lib-agent-cpp/inc/principal.h [hunk 4]
```diff
@@ -32,31 +32,123 @@ class Principal {
   std::vector<unsigned char> bytes;
   CPrincipal *cPrincipal;
 
-  static void error_callback(const unsigned char *data, int len,
+/**
```
Comment (neithanmo) [inline lib-agent-cpp/inc/principal.h:35]: We do not need to document private functions, this callback I think is not public, and should not be, it is used internally to handle error returns

**Pull request merged**